src/lib/operators/export_binary.cpp

```diff
@@ -171,90 +149,70 @@ void ExportBinary::_write_header(const Table& table, std::ofstream& ofstream) {
   }
 }
 
-template <typename T>
-void ExportBinary::ExportBinaryVisitor<T>::handle_segment(const BaseValueSegment& base_segment,
-                                                          std::shared_ptr<SegmentVisitorContext> base_context) {
-  auto context = std::static_pointer_cast<ExportContext>(base_context);
-  const auto& segment = static_cast<const ValueSegment<T>&>(base_segment);
+void ExportBinary::_write_segment(const BaseEncodedSegment& base_segment, std::ofstream& ofstream) {
+  Fail("Binary export for segment type is not supported yet.");
+}
 
-  export_value(context->ofstream, BinarySegmentType::value_segment);
+template <typename T>
+void ExportBinary::_write_segment(const ValueSegment<T>& segment, std::ofstream& ofstream) {
+  export_value(ofstream, EncodingType::Unencoded);
 
   if (segment.is_nullable()) {
-    export_values(context->ofstream, segment.null_values());
+    export_values(ofstream, segment.null_values());
   }
 
-  export_values(context->ofstream, segment.values());
+  export_values(ofstream, segment.values());
 }
 
-template <typename T>
-void ExportBinary::ExportBinaryVisitor<T>::handle_segment(const ReferenceSegment& ref_segment,
-                                                          std::shared_ptr<SegmentVisitorContext> base_context) {
-  auto context = std::static_pointer_cast<ExportContext>(base_context);
-
+void ExportBinary::_write_segment(const ReferenceSegment& ref_segment, std::ofstream& ofstream) {
   // We materialize reference segments and save them as value segments
-  export_value(context->ofstream, BinarySegmentType::value_segment);
+  export_value(ofstream, EncodingType::Unencoded);
 
-  // Unfortunately, we have to iterate over all values of the reference segment
-  // to materialize its contents. Then we can write them to the file
-  for (ChunkOffset row = 0; row < ref_segment.size(); ++row) {
-    export_value(context->ofstream, boost::get<T>(ref_segment[row]));
-  }
-}
+  if (ref_segment.data_type() == DataType::String) {
+    if (ref_segment.size() == 0) return;
 
-// handle_segment implementation for string segments
-template <>
-void ExportBinary::ExportBinaryVisitor<pmr_string>::handle_segment(
-    const ReferenceSegment& ref_segment, std::shared_ptr<SegmentVisitorContext> base_context) {
-  auto context = std::static_pointer_cast<ExportContext>(base_context);
+    std::stringstream values;
+    pmr_string value;
+    pmr_vector<size_t> string_lengths(ref_segment.size());
 
-  // We materialize reference segments and save them as value segments
-  export_value(context->ofstream, BinarySegmentType::value_segment);
-
-  // If there is no data, we can skip all of the coming steps.
-  if (ref_segment.size() == 0) return;
+    // We export the values materialized
+    for (ChunkOffset row = 0; row < ref_segment.size(); ++row) {
+      value = boost::get<pmr_string>(ref_segment[row]);
```

Comment:
No need to change this as we are not yet looking at the performance of the export, but `operator[]` is quite slow because it converts everything into an AllTypeVariant. For new code, have a look at `segment_accessor.hpp`.